### PR TITLE
fix(vitest): strip ansi code from console output in junit reporter

### LINKED ISSUE #4958: JUnit reporter sometimes adds ansi colors to the report
### Describe the bug

Hey, from time to time the JUnit reporter adds ansi colors to my report (not sure what's the rule), and this makes the report unreadable in places like the CircleCI's test results which doesn't show ansi colors.

This is how it looks like:
image

I used to have a `NO_COLOR`/`--no-color` flag, but it applies to both reporters
How can I configure the junit reporter to strip the colors?

### Reproduction

I will add a stackblitz once I manage to reproduce it in a standalone environment.
Right now I am experiencing this in my local machine.
I'll add a reproduction once I manage to create one 💪🏼 

### System Info

```shell
System:
    OS: macOS 14.2.1
    CPU: (12) x64 Apple M2 Max
    Memory: 23.80 MB / 32.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 20.10.0 - ~/.volta/tools/image/node/20.10.0/bin/node
    Yarn: 1.22.19 - ~/.volta/tools/image/yarn/1.22.19/bin/yarn
    npm: 10.2.3 - ~/.volta/tools/image/node/20.10.0/bin/npm
  Browsers:
    Chrome: 120.0.6099.216
    Safari: 17.2.1
```

### Used Package Manager

yarn

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

### PR DESCRIPTION
**Pull request by hi-ogawa**
### Description

- Closes https://github.com/vitest-dev/vitest/issues/4958

I'm not sure the need of preserving color for this logging output, so I propose to always strip it for junit without configuration to toggle color. Please let me know if this makes sense.

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [x] Ideally, include a test that fails without this PR but passes with it.
- [x] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.

### Tests
- [ ] Run the tests with `pnpm test:ci`.

### Documentation
- [ ] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [x] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

### CODE REVIEW

**Comment by AriPerkkio** [comment]
I remember adding this environment variable for reporters test case just because of this issue. Can we now remove it from here?

https://github.com/vitest-dev/vitest/blob/a4ae57440aaa59b858b4e00a14e31ffd40709291/test/reporters/package.json#L6

File: test/reporters/tests/html.test.ts [hunk 1]
```diff
@@ -47,7 +47,9 @@ describe('html reporter', async () => {
     const metaJsonGzipped = fs.readFileSync(resolve(root, `${basePath}/html.meta.json.gz`))
     const metaJson = zlib.gunzipSync(metaJsonGzipped).toString('utf-8')
     const indexHtml = fs.readFileSync(resolve(root, `${basePath}/index.html`), { encoding: 'utf-8' })
-    const resultJson = parse(metaJson.replace(new RegExp(vitestRoot, 'g'), '<rootDir>'))
+    const resultJson = parse(metaJson.replace(new RegExp(vitestRoot, 'g'), '<rootDir>'), (_key, value) => {
+      return typeof value === 'string' ? stripAnsi(value) : value
+    })
```
Comment (hi-ogawa) [inline test/reporters/tests/html.test.ts:52]: `NO_COLOR` was also affecting html reporter, so I stripped it just for this test case.

**Comment by sheremet-va** [comment]
I am not sure why this is a problem. `NO_COLOR` disables this, right? What makes this reporter special that we always have to remove the colors?

**Comment by AriPerkkio** [comment]
> What makes this reporter special that we always have to remove the colors?

I think it makes sense for users to keep colors enabled for reporters that output on terminal on CI. But they would still like to generate other reports on file system and keep their content human-readable.

**Comment by hi-ogawa** [comment]
> I think it makes sense for users to keep colors enabled for reporters that output on terminal on CI. But they would still like to generate other reports on file system and keep their content human-readable.

I haven't properly followed up with the issue author https://github.com/vitest-dev/vitest/issues/4958 but I assumed that's the intention and made sense to me as a default junit reporter behavior. But, disabling entirely by `NO_COLOR` should be not a big deal and I totally get that applying `stripAnsi` here feels somewhat too artificial.

**Comment by sheremet-va** [comment]
> I think it makes sense for users to keep colors enabled for reporters that output on terminal on CI. But they would still like to generate other reports on file system and keep their content human-readable.

It just looks too irreversible to me. As @hi-ogawa, what if the color is expected? I think we can add an option to toggle this. We can even disable colors by default if you think it is a better experience for most people.

**Comment by hi-ogawa** [comment]
Thanks for the discussion. Yeah, I'll re-consider this change. Also, it's better to wait for the issue author's response to know the actual experience with current junit output and measure the significance of the issue. I will put this back to draft.

**Comment by AriPerkkio** [comment]
Another approach for this would be support for reporter options. I would imagine something like: 

```ts
{
  test: {
    reporters: ["verbose", ["junit", { color: false }], "html"],
  },
}
```

This was mentioned here https://github.com/vitest-dev/vitest/issues/3691#issuecomment-1611162791 but no actions were taken. This would also benefit [third party reporters](https://github.com/AriPerkkio/vitest-sonar-reporter?tab=readme-ov-file#options).

**Comment by sheremet-va** [comment]
> Another approach for this would be support for reporter options

I think that would be nice to support but we didn't create an issue yet, true.

**Comment by hi-ogawa** [comment]
That sounds good to me. I created an issue here https://github.com/vitest-dev/vitest/issues/5042

**Comment by AriPerkkio** [comment]
Now that #5042 is finished, the JUnit reporter accepts some reporter options:

https://github.com/vitest-dev/vitest/blob/5ed537f03ea33df3516e530b55a5c1f0fc219e24/packages/vitest/src/node/reporters/junit.ts#L15-L19

Maybe a new option should be introduced there, `color?: boolean` that defaults to `true`. Then user's could disable it like shown in https://github.com/vitest-dev/vitest/pull/5007#issuecomment-1902094894.

**Comment by hi-ogawa** [comment]
@AriPerkkio Thanks for the ping. For this junit issue, I'm becoming less convinced about this feature, so I'll close this for now. I think we can reconsider this if we get more reports.

**Comment by ssbarnea** [comment]
The original bug is still true and using environment variables is not possible as they have other side effects, like disabling coloring on vitest itself and worse, affecting the testing, as the tool being tested might react to the NO_COLOR, including node itself!

Sadly the original bug was closed and locked.

**Pull request closed without merging**